LUCENE-9390: JapaneseTokenizer discards token that is all punctuation characters only

johtani opened the pull request:
# Description

Check and omit token that has all punctuation characters when discard punctuation flag is true.
Currently, JapaneseTokenizer discards token that has punctuation at first character only.

# Solution

Add isAllPunctuation method for testing token.

# Tests

Ensure to discard if token is all punctuation characters.
And not discard if token that start punctuation character and has non-punctuation character.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

jimczi commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/JapaneseTokenizer.java:
```diff
@@ -1917,4 +1917,15 @@ private static boolean isPunctuation(char ch) {
+    boolean flag = true;
+    for (int i = offset; i < offset + length; i++) {
+      if (!isPunctuation(ch[i])) {
+        flag = false;
```
nit: you can return `false` directly ?

johtani replied:
Fixed this.

jimczi commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/JapaneseTokenizer.java:
```diff
@@ -1917,4 +1917,15 @@ private static boolean isPunctuation(char ch) {
+        break;
+      }
+    }
+    return flag;
```
return `true` ?

johtani replied:
Fixed this.

johtani commented:
I added NBest test case. And also I changed registerNode. 
However, there is no difference between changing it or not changing it...
Am I missing something test case?

For NBest test case with discard punctuation, the tokenizer outputs a complicated token stream, so [I set `graphOffsetsAreCorrect` is `false`](https://github.com/apache/lucene-solr/blob/abf243c5cec331ec8419f0fd7c966dbce45f6b2d/lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/TestJapaneseTokenizer.java#L967).
